### ISSUE TITLE
When CWND is low, the pacer signals its user to pause more often

### ISSUE THREAD

**Issue by KershawChang** [mozilla/neqo#1463]
I've been debugging the upload performance issue, and it seems that the bottleneck is related to the pacer.
According to this equation ["(now - self.t) * PACER_SPEEDUP * cwnd / rtt"](https://searchfox.org/mozilla-central/rev/118283c2e3eb393e9b1ca00a0c328e32eeab1c40/third_party/rust/neqo-transport/src/pace.rs#83), when the `cwnd` is small, the pacer increase its capability with a smaller value.
As the following log shows, there is a `102.471µs` delay after sending a few packets. Unfortunately, the [timer resolution in necko is in milliseconds](https://searchfox.org/mozilla-central/rev/118283c2e3eb393e9b1ca00a0c328e32eeab1c40/netwerk/socket/neqo_glue/src/lib.rs#329), this delay becomes 1ms. 
```
2023-09-27 20:54:04.019175 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 728/1357] next 121081/31.174185ms wait 80.972µs = Instant { tv_sec: 3158011, tv_nsec: 382555722 }
2023-09-27 20:54:04.019178 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::connection] [Client a228a8261e9e8b4596] output_path send_profile SendProfile { limit: 1357, pto: None, probe: (), paced: false }
2023-09-27 20:54:04.019189 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 728/1357] spend 1357 cwnd=[121081], rtt=[31.174185ms] t=1.295291ms incr=10061
2023-09-27 20:54:04.019198 UTC - [Parent 70649: Socket Thread]: E/nsHttp Http3Session::ProcessOutput sending packet with 1357 bytes to 142.250.181.238 port=443 [this=1403e1f00].
2023-09-27 20:54:04.019214 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 2714/1357] next 121081/31.174185ms no wait = Instant { tv_sec: 3158011, tv_nsec: 383770041 }
2023-09-27 20:54:04.019217 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::connection] [Client a228a8261e9e8b4596] output_path send_profile SendProfile { limit: 1357, pto: None, probe: (), paced: false }
2023-09-27 20:54:04.019226 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 2714/1357] spend 1357 cwnd=[121081], rtt=[31.174185ms] t=38.792µs incr=301
2023-09-27 20:54:04.019234 UTC - [Parent 70649: Socket Thread]: E/nsHttp Http3Session::ProcessOutput sending packet with 1357 bytes to 142.250.181.238 port=443 [this=1403e1f00].
2023-09-27 20:54:04.019247 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 1658/1357] next 121081/31.174185ms no wait = Instant { tv_sec: 3158011, tv_nsec: 383808833 }
2023-09-27 20:54:04.019250 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::connection] [Client a228a8261e9e8b4596] output_path send_profile SendProfile { limit: 1357, pto: None, probe: (), paced: false }
2023-09-27 20:54:04.019259 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 1658/1357] spend 1357 cwnd=[121081], rtt=[31.174185ms] t=33.5µs incr=260
2023-09-27 20:54:04.019267 UTC - [Parent 70649: Socket Thread]: E/nsHttp Http3Session::ProcessOutput sending packet with 1357 bytes to 142.250.181.238 port=443 [this=1403e1f00].
2023-09-27 20:54:04.019279 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 561/1357] next 121081/31.174185ms wait 102.471µs = Instant { tv_sec: 3158011, tv_nsec: 383944804 }
2023-09-27 20:54:04.019282 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::connection] [Client a228a8261e9e8b4596] output_path send_profile SendProfile { limit: 255, pto: None, probe: (), paced: true }
2023-09-27 20:54:04.019286 UTC - [Parent 70649: Socket Thread]: I/neqo_transport::* [neqo_transport::pace] [Pacer id=7 561/1357] next 121081/31.174185ms wait 102.471µs = Instant { tv_sec: 3158011, tv_nsec: 383944804 }
2023-09-27 20:54:04.019289 UTC - [Parent 70649: Socket Thread]: E/nsHttp Http3Session::SetupTimer to 1ms [this=1403e1f00].
```

Moreover, when necko/neqo is stuck in this state, it's almost not possible to recover. This is because the acked bytes is lower than [bytes_for_increase](https://searchfox.org/mozilla-central/rev/118283c2e3eb393e9b1ca00a0c328e32eeab1c40/third_party/rust/neqo-transport/src/cc/classic_cc.rs#206), so the cwnd has no chance to grow.

@jesup , @martinthomson , could you take a look and see if my finding makes sense?
Thanks.

**Comment by KershawChang** [comment]
So, I attempted to completely disable the pacer, yet we remain unable to recover from the state where `cwnd` remains low. This leads me to think the core issue may be around the reason `cwnd` stays at a low value. I've attached an image below that shows the evolution of `cwnd` over time. From my observations, it appears there are two main contributing factors preventing `cwnd` from increasing:

1. The flat line observed around 21:02:15 is, I think, a result of app_limit. For approximately 5 seconds, cwnd isn't allowed to increase due to app_limited.

2. After 21:02:15, `cwnd` appears to be increasing in a linear fashion, consistent with the behavior defined by the New Reno algorithm. However, the growth doesn't seem as effective as anticipated. Not sure if this is related to the pacer or not.

![new_reno](https://github.com/mozilla/neqo/assets/6964789/6ef07516-c496-4cd2-8501-021ad4e686b6)

**Comment by martinthomson** [comment]
That looks like a very lossy connection.  I'm seeing lots of drops.  Is there any way to track places where packets were marked as lost?

**Comment by KershawChang** [comment]
> That looks like a very lossy connection. I'm seeing lots of drops. Is there any way to track places where packets were marked as lost?

Maybe the best way is using qlog, so I'll use qlog to share our findings in the future.

I've recorded a qlog [here](https://drive.google.com/file/d/1-6_KAypxC_VNpCCg5Odw1zfwM48ptboj/view?usp=drive_link).
As the screen shot below shows, we might have two problems:
1. The app_limit seems to prevent us from increasing cwnd after a packet loss.
2. For some reason, the max `bytes_in_flight` is only half of the `cwnd`. I assume the `bytes_in_flight` should be closed to `cwnd` as much as possible?
<img width="2299" alt="Screenshot 2023-10-05 at 3 54 45 PM" src="https://github.com/mozilla/neqo/assets/6964789/c16fe4f5-72cd-4633-8bc5-b3b1a215e8e2">

**Comment by KershawChang** [comment]
FYI, these are two more qlogs from Andrew.

- New Reno
<img width="2302" alt="Screenshot 2023-10-05 at 9 19 25 PM" src="https://github.com/mozilla/neqo/assets/6964789/5f5d1377-86c1-4f54-a11b-27705be72c2f">

- Cubic
<img width="2302" alt="Screenshot 2023-10-05 at 9 22 04 PM" src="https://github.com/mozilla/neqo/assets/6964789/1cb59869-aa69-4580-b756-bdd39230a68b">

Apparently, in these two graphs, the `bytes_in_flight` can reach the full value of `cwnd` sometimes. I'm not sure what could cause this difference, but I'll try to figure it out.

**Comment by martinthomson** [comment]
Try disabling the app limited cap as well and see what happens there.

**Comment by junhochoi** [comment]
From above 3 qlogs, I wonder why cwnd increases when bytes_flights is much lower than cwnd. It looks like a violation of https://datatracker.ietf.org/doc/html/rfc9002#section-7.8, assuming no flow control limited. Doesn't look like a problem of a specific CCA.

**Comment by junhochoi** [comment]
For example, in "reno" graph, why cwd increases around 36,255ms (apparently, bytes_in_flights is far lower than cwnd, looks like an app limited)?

**Comment by KershawChang** [comment]
> Try disabling the app limited cap as well and see what happens there.

I've tried disabling the app_limited, but doing so by itself doesn't improve performance. As noted in the first comment, the pacer still slows down throughput.

I found that adjusting the following parameters can help:
- `PACING_BURST_SIZE`: This not only increases the pacer's maximum capacity but also reduces the occurrences of app_limited.
- `PACER_SPEEDUP`: This allocates more credits to the pacer when the `spend` function is called.
- `SEND_BUFFER_SIZE`: It's currently set to 1MB, but I'm unsure if this is enough. Increasing this value allows Necko to fill more data, leading to improved throughput.

Can we make these parameters configurable? This would enable us to fine-tune them through experimentation.
